android/src/com/frostwire/android/gui/fragments/preference/ApplicationFragment.java

```diff
@@ -70,9 +72,39 @@ protected void initComponents() {
         setupConnectSwitch();
         setupVPNRequirementOption();
         setupStorageOption();
+        setupDataSaving();
         setupStore(removeAdsPurchaseTime);
     }
 
+    private void setupDataSaving() {
+        SwitchPreferenceCompat preference = findPreference(Constants.PREF_KEY_NETWORK_USE_WIFI_ONLY);
+        preference.setOnPreferenceChangeListener(new Preference.OnPreferenceChangeListener() {
+            @Override
+            public boolean onPreferenceChange(final Preference preference, Object newValue) {
+                boolean newVal = (Boolean) newValue;
+                if (newVal && !NetworkManager.instance().isDataWIFIUp()) {
+                    if (TransferManager.instance().isHttpDownloadInProgress()) {
+                        UIUtils.showYesNoDialog(getActivity(), R.string.data_saving_kill_http_warning, R.string.data_saving_kill_http_warning_title, new DialogInterface.OnClickListener() {
```

Comment:
I don't think this `UIUtils.showYesNoDialog` is safe anymore. Did you test rotation and context leaks? Take a look at `KitKatStoragePreference` for an example of how to use `YesNoDialog`